pkg/apis/discovery/v1beta1/conversion.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/discovery/v1beta1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/kubernetes/pkg/apis/discovery"
+)
+
+func Convert_v1beta1_Endpoint_To_discovery_Endpoint(in *v1beta1.Endpoint, out *discovery.Endpoint, s conversion.Scope) error {
+	if err := autoConvert_v1beta1_Endpoint_To_discovery_Endpoint(in, out, s); err != nil {
+		return err
+	}
+
+	// Copy	Topology into Deprecated Topology
+	out.DeprecatedTopology = make(map[string]string, len(in.Topology))
+	for k, v := range in.Topology {
+		out.DeprecatedTopology[k] = v
+	}
+
+	// Move zone from the topology map into a field
+	if zone, ok := in.Topology[corev1.LabelTopologyZone]; ok {
+		out.Zone = &zone
+		delete(out.DeprecatedTopology, corev1.LabelTopologyZone)
+	}
+
+	// If zone was the only field in the map or topology was empty set DeprecatedTopology
+	// to nil
+	if len(out.DeprecatedTopology) == 0 {
+		out.DeprecatedTopology = nil
+	}
+
+	return nil
+}
+
+func Convert_discovery_Endpoint_To_v1beta1_Endpoint(in *discovery.Endpoint, out *v1beta1.Endpoint, s conversion.Scope) error {
```

Comment:
I thought when converting to v1, we were going to drop the hostname label from DeprecatedTopology if it matched nodeName, so under typical circumstances, a v1 client would never see the deprecatedTopology field at all

Reply:
I think we agreed to leave it alone - if the user wrote 2 different values they get to keep them, unless they write thru v1.  I don't see any real reason to unpack it further?

Reply:
the propagation of nodeName → deprecatedTopology[hostname] when writing via v1 is what I'm trying to think through... won't that make `deprecatedTopology[hostname]` always be populated in v1? I thought we wanted that empty/nil/omitted in the happy "write via v1 and then read via v1" path

Reply:
You're right that `deprecatedTopology[hostname]` should show up in v1 but it doesn't (in my branch).  Hmm.  Will have to try this branch tomorrow.

Reply:
OK, I confirm that this DOES set `deprecatedTopology[hostname]`, as expected.  The question is whether we care?

I'll see what a quick fix woould look like

Reply:
This seems to work - Swetha, what do you think?  I did not update tests, but manual testing seems to work.   I think this solves the other issue too?

```
diff --git a/pkg/apis/discovery/v1beta1/conversion.go b/pkg/apis/discovery/v1beta1/conversion.go
index f741c5955c8..507a28afcbe 100644
--- a/pkg/apis/discovery/v1beta1/conversion.go
+++ b/pkg/apis/discovery/v1beta1/conversion.go
@@ -40,6 +40,16 @@ func Convert_v1beta1_Endpoint_To_discovery_Endpoint(in *v1beta1.Endpoint, out *d
 		delete(out.DeprecatedTopology, corev1.LabelTopologyZone)
 	}
 
+	// Remove hostname from the topology map ONLY IF it is the same value as
+	// nodeName.  This preserves the (rather odd) ability to have different
+	// values for topology[hostname] and nodename in v1beta1, without showing
+	// duplicate values in v1.
+	if node, ok := in.Topology[corev1.LabelHostname]; ok {
+		if out.NodeName != nil && node == *out.NodeName {
+			delete(out.DeprecatedTopology, corev1.LabelHostname)
+		}
+	}
+
 	// If zone was the only field in the map or topology was empty set DeprecatedTopology
 	// to nil
 	if len(out.DeprecatedTopology) == 0 {
@@ -55,7 +65,7 @@ func Convert_discovery_Endpoint_To_v1beta1_Endpoint(in *discovery.Endpoint, out
 	}
 
 	// If no deprecated topology or zone field, no conversion is necessary
-	if in.DeprecatedTopology == nil && in.Zone == nil {
+	if in.DeprecatedTopology == nil && in.Zone == nil && in.NodeName == nil {
 		return nil
 	}
 
@@ -70,5 +80,12 @@ func Convert_discovery_Endpoint_To_v1beta1_Endpoint(in *discovery.Endpoint, out
 		out.Topology[corev1.LabelTopologyZone] = *in.Zone
 	}
 
+	// Add hostname into the topology map ONLY IF it is not already present.
+	// This preserves the (rather odd) ability to have different values for
+	// topology[hostname] and nodename in v1beta1.
+	if in.NodeName != nil && out.Topology[corev1.LabelHostname] == "" {
+		out.Topology[corev1.LabelHostname] = *in.NodeName
+	}
+
 	return nil
 }
diff --git a/pkg/registry/discovery/endpointslice/strategy.go b/pkg/registry/discovery/endpointslice/strategy.go
index d8be219fd07..bf287afbf2a 100644
--- a/pkg/registry/discovery/endpointslice/strategy.go
+++ b/pkg/registry/discovery/endpointslice/strategy.go
@@ -19,7 +19,6 @@ package endpointslice
 import (
 	"context"
 
-	corev1 "k8s.io/api/core/v1"
 	discoveryv1 "k8s.io/api/discovery/v1"
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -146,8 +145,7 @@ func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
 	}
 }
 
-// dropTopologyOnV1 on V1 request wipes the DeprecatedTopology field  and copies
-// the NodeName value into DeprecatedTopology
+// dropTopologyOnV1 on V1 request wipes the DeprecatedTopology field.
 func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSlice) {
 	if info, ok := genericapirequest.RequestInfoFrom(ctx); ok {
 
@@ -167,14 +165,6 @@ func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSli
 					//Silently clear out DeprecatedTopology
 					ep.DeprecatedTopology = nil
 				}
-
-				// Set topology[hostname] to match NodeName for compatibility with old clients
-				if ep.NodeName != nil {
-					if ep.DeprecatedTopology == nil {
-						ep.DeprecatedTopology = make(map[string]string, 1)
-					}
-					ep.DeprecatedTopology[corev1.LabelHostname] = *ep.NodeName
-				}
 			}
 		}
 	}
```

Reply:
If in v1beta1, `nodeName` is not populated, should we sync that with `topology[hostname]`?  Otherwise  v1beta1 -> v1, only deprecatedTopology has the hostname.

Reply:
since nodeName is a field in v1beta1, I think it's easier to just leave it mapped to the v1 field.  read-clients can fall-back on topology, but when v1beta1 goes away, write-clients need to update to the field anyway.

Reply:
The only other consequence I see of this approach is that if in v1beta1 only `NodeName` is set, after the round trip, `topology[hostname]` will also be set. I don't think that is too much of an issue and it is behavior that will go away once v1beta1 is removed.